File: Engine/Source/Scene/Scene.cpp
#include "Scene.h"
#include "Utilities\Utilities.h"

#include "Core\Engine.h"
#include "Graphics\Interface\IRenderer.h"
#include "Graphics\Material.h"

#include "Graphics\Primitives.h"

namespace Scene
{
	bool Scene::Initialize()
	{
		auto ptr = InstantiateGameObject("Camera");

		auto camera = std::make_unique<CCamera>();
		camera->Initialize(ptr, 70.0f, 1, 5000.0f);

		ptr->AddComponent(std::move(camera));

		auto script = std::make_unique<CScriptComponent>();
		script->Initialize(ptr, "Assets/Scripts/Camera.lua");

		ptr->AddComponent(std::move(script));

		auto trans = std::make_unique<Transformation>();
		trans->Position = Vector3f(0, 0, 0);
		trans->Scale = Vector3f(15, 15, 15);

		ptr = InstantiateGameObject("Water", std::move(trans));

		auto mesh = std::make_unique<CMeshRenderer>();

		auto tempMat = std::make_unique<Graphics::Material>();
		tempMat->TextureFilter = Graphics::ETextureFilter::NEAREST;

		Graphics::STextureDesc texInfo;
		texInfo.RegisterIndex = 0;
		texInfo.UniformName = "uNoise";
		texInfo.FilePath = "Assets/Textures/Noise.tga";
		tempMat->AddTexture(texInfo);

		texInfo.RegisterIndex = 1;
		texInfo.UniformName = "uColor";
		texInfo.FilePath = "Assets/Textures/Colormap.tga";
		tempMat->AddTexture(texInfo);

		tempMat->Shaders.VertexShaderPath = "Terrain.vs";
		tempMat->Shaders.FragmentShaderPath = "Terrain.fs";
		tempMat->Shaders.ShaderContainerName = "Terrain";
		
		Graphics::SMeshDesc desc{};
		Graphics::Primitives::GetPlaneTri(desc, 100, 0, 100);
		desc.Mode = Graphics::EMeshPrimitive::TRIANGLES;
		desc.HasIndices = true;
		desc.FilePath = "Primitive";

		mesh->Initialize(ptr, desc, std::move(tempMat));
		ptr->AddComponent(std::move(mesh));

		trans = std::make_unique<Transformation>();
		trans->Position = Vector3f(10, -10, 0);
		trans->Scale = Vector3f(1, 1, 1);		

		ptr = InstantiateGameObject("Terrain", std::move(trans));

		mesh = std::make_unique<CMeshRenderer>();

		tempMat = std::make_unique<Graphics::Material>();
		tempMat->TextureFilter = Graphics::ETextureFilter::NEAREST;

		tempMat->Shaders.VertexShaderPath = "Test.vs";
		tempMat->Shaders.FragmentShaderPath = "Test.fs";
		tempMat->Shaders.ShaderContainerName = "Test";

		ZeroMemory(&desc, sizeof(Graphics::SMeshDesc));
		desc.HasIndices = true;
		desc.Mode = Graphics::EMeshPrimitive::TRIANGLES;
		desc.FilePath = "Assets/Models/teapot.obj";

		mesh->Initialize(ptr, desc, std::move(tempMat));
		ptr->AddComponent(std::move(mesh));

		script = std::make_unique<CScriptComponent>();
		script->Initialize(ptr, "Assets/Scripts/Test.lua");

		ptr->AddComponent(std::move(script));

		return true;
	}

	void Scene::Update()
	{
		for (auto &temp : m_GameObjects)
		{
			temp.second->Update();
		}
	}

	Scene::~Scene()
	{
		ClearScene();
	}

	bool Scene::AddGameObject(std::unique_ptr<GameObject> a_ToAdd)
	{
		if (a_ToAdd == nullptr)
		{
			LogErr("AddGameObject() GameObject is Null\n");
			return false;
		}
		
		if (m_GameObjects.find(a_ToAdd->Name) != m_GameObjects.end())
		{
			LogErr("GameObject [%s] is already added\n", a_ToAdd->Name.c_str());
			a_ToAdd = nullptr;
			return false;
		}

		LogErr("GameObject added %s\n", a_ToAdd->Name.c_str());
		m_GameObjects[a_ToAdd->Name] = std::move( a_ToAdd );
		return true;
	}

	void Scene::AddRenderable(GameObject *a_Parent, CMeshRenderer *a_Renderable)
	{
		if (CheckIfPointerIsInvalid(a_Parent))
		{
			LogErr("AddRenderable() GameObject is Null\n");
			return;
		}

		if (CheckIfPointerIsInvalid(a_Renderable))
		{
			LogErr("AddRenderable() Renderable is Null\n");
			return;
		}

		//Already added
		if (m_Renderables.find(a_Parent) != m_Renderables.end())
		{
			LogErr("AddRenderable() GameObject already added\n");
			return;
		}
		else
		{
			m_Renderables.insert(std::pair<GameObject*, CMeshRenderer*>(a_Parent, a_Renderable));
			Core::Engine::StaticClass()->GetRenderer()->AddRenderable(a_Renderable);
		}
	}

	void Scene::RemoveRenderable(GameObject *a_Parent)
	{
		if (CheckIfPointerIsInvalid(a_Parent))
		{
			LogErr("RemoveRenderable() GameObject is Null\n");
			return;
		}

		if (m_Renderables.find(a_Parent) != m_Renderables.end())
		{
			CMeshRenderer *temp = m_Renderables.find(a_Parent)->second;			
			//The Renderer need to notice the deletion			
			Core::Engine::StaticClass()->GetRenderer()->RemoveRenderable(temp);
			m_Renderables.erase(a_Parent);
		}
		else
		{
			LogErr("RemoveRenderable() GameObject not found\n");
			return;
		}
	}

	void Scene::RemoveGameObject(std::string &a_Name)
	{
		if (m_GameObjects.find(a_Name) != m_GameObjects.end())
		{
			m_GameObjects.erase(a_Name);
		}
		else
		{
			LogErr("RemoveGameObject() GameObject does not exist\n");
		}
	}

	GameObject* Scene::InstantiateGameObject(std::string a_Name)
	{
		return InstantiateGameObject(a_Name, std::make_unique<Transformation>());
	}

	GameObject* Scene::InstantiateGameObject(std::string a_Name, std::unique_ptr<Transformation> a_Transform)
	{
		auto temp = std::make_unique<GameObject>();
		temp->Name = a_Name;

		auto ptr = temp.get();

		if (!AddGameObject(std::move(temp)))
			return nullptr;

		ptr->SetTransform(std::move(a_Transform));

		return ptr;
	}

	void Scene::ClearScene()
	{
		m_GameObjects.clear();
		m_Renderables.clear();
	}

	void Scene::SetCamera(CCamera *a_Camera)
	{	m_Camera = a_Camera;	}
	
	CCamera* Scene::GetCamera()
	{	return m_Camera;	}

	std::unordered_map<GameObject*, CMeshRenderer*> Scene::GetRenderables()
	{	return m_Renderables;	}

}

File: Sandbox/Assets/Scripts/Scene.lua
index = 0

function Init()
	-- Camera Object
	trans = Transform.new();
	trans.Rotation = Vector3f.new(-40, 40, 0)
	
	gameObject = Scene:InstantiateGameObject("Camera", trans, true)
	
	script = gameObject:AddScriptComponent()
	script:Initialize("Assets/Scripts/Camera.lua")
	
	camera = gameObject:AddCameraComponent();
	camera:Initialize(80.0, 1, 5000.0)
	
	--image = Image.new("Assets/Textures/TreeMap.tga");	
	--noiseImage = Image.new("Assets/Textures/Noise.tga");	
	
	-- Seating
	
	trans.Position = Vector3f.new(0, 14.5, 0)
	trans.Rotation = Vector3f.new(2,0,-2)
	trans.Scale = Vector3f.new(0.1, 0.1, 0.1);
	
	gameObject = Scene:InstantiateGameObject("Seating", trans, false)
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(0, "Small.vs")
	material:AddShader(0, "Small.fs")
	material:AddShader(1, "Shadow/Depth.vs")
	material:AddShader(1, "Shadow/Depth.fs")
	
	texDesc = TextureDesc.new()
	texDesc.RegisterIndex = 0
	texDesc.UniformName = "uTexture"
	texDesc.FilePath = "Assets/Textures/Diffuse.tga"
	material:AddTexture(texDesc)
	
	mesh:Initialize("Assets/Models/Seating.obj", material)
	
	-- Boat
	
	trans.Position = Vector3f.new(850, 0.5, 0)
	trans.Rotation = Vector3f.new(0, 30, 0)
	trans.Scale = Vector3f.new(0.2, 0.2, 0.2);
	
	gameObject = Scene:InstantiateGameObject("Boat", trans, true)
	
	script = gameObject:AddScriptComponent()
	script:Initialize("Assets/Scripts/Boat.lua")
	
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(0, "Small.vs")
	material:AddShader(0, "Small.fs")
	material:AddShader(1, "Shadow/Depth.vs")
	material:AddShader(1, "Shadow/Depth.fs")
	
	texDesc = TextureDesc.new()
	texDesc.RegisterIndex = 0
	texDesc.UniformName = "uTexture"
	texDesc.FilePath = "Assets/Textures/Diffuse.tga"
	material:AddTexture(texDesc)
	
	mesh:Initialize("Assets/Models/Boat.obj", material)
	
	-- Planks
	
	trans.Position = Vector3f.new(790, 0, 0)
	trans.Scale = Vector3f.new(0.2, 0.2, 0.2);
	trans.Rotation = Vector3f.new(0, 0, 0)
	
	gameObject = Scene:InstantiateGameObject("Plank", trans, false)
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(0, "Small.vs")
	material:AddShader(0, "Small.fs")
	material:AddShader(1, "Shadow/Depth.vs")
	material:AddShader(1, "Shadow/Depth.fs")
	
	texDesc = TextureDesc.new()
	texDesc.RegisterIndex = 0
	texDesc.UniformName = "uTexture"
	texDesc.FilePath = "Assets/Textures/Diffuse.tga"
	material:AddTexture(texDesc)
	
	mesh:Initialize("Assets/Models/Plank.obj", material)
	
	trans.Rotation = Vector3f.new(0, 0, 0)
	for index=0, 500, 1 do
	
		scale = math.random() * 0.25 + 0.2
		trans.Rotation = Vector3f.new(math.random(-2, 2), math.random(0, 360), math.random(-2, 2))
		trans.Scale = Vector3f.new(scale, scale, scale);
		trans.Position = Vector3f.new(math.random(0, 2000), 35, math.random(0, 2000))
		
		--if image:GetPixelAt(trans.Position.x, trans.Position.z).x > 250 then
			
			--trans.Position.y = (noiseImage:GetPixelAt(trans.Position.x, trans.Position.z).x /255) * 50 - 26
			trans.Position.y = 15
			trans.Position.x = trans.Position.x - 1024
			trans.Position.z = trans.Position.z - 1024
			
			gameObject = Scene:InstantiateGameObject(tostring(index), trans, false)
			mesh = gameObject:AddMeshRenderer()
			material = mesh:SetNewMaterial()
			
			material:AddShader(0, "Tree.vs")
			material:AddShader(0, "Tree.fs")
			
			material:AddShader(1, "Shadow/DepthTree.vs")
			material:AddShader(1, "Shadow/DepthTree.fs")
			
			texDesc.RegisterIndex = 0
			texDesc.UniformName = "uTexture"
			texDesc.FilePath = "Assets/Textures/Diffuse.tga"
			material:AddTexture(texDesc)
			
			texDesc.RegisterIndex = 1
			texDesc.UniformName = "uLeaves"
			texDesc.FilePath = "Assets/Textures/Leaves.tga"
			material:AddTexture(texDesc)
			
			mesh:Initialize("Assets/Models/Tree.obj", material)
			
		--end
	end 
	
	-- Terrain	
	texDesc = TextureDesc.new()
	meshDesc = MeshDesc.new()
	
	trans = Transform.new()
	trans.Scale = Vector3f.new(15, 15, 15)
	
	gameObject = Scene:InstantiateGameObject("Terrain", trans, false)
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(0, "Terrain.vs")
	material:AddShader(0, "Terrain.fs")
	
	texDesc.RegisterIndex = 0
	texDesc.UniformName = "uNoise"
	texDesc.FilePath = "Assets/Textures/Noise.tga"
	material:AddTexture(texDesc)
	
	texDesc.RegisterIndex = 1
	texDesc.UniformName = "uColor"
	texDesc.FilePath = "Assets/Textures/Colormap.tga"
	material:AddTexture(texDesc)
	
	-- Primitives is static
	Primitives.GetPlaneTri(meshDesc, 150, 0, 150)
	meshDesc.HasIndices = true
	meshDesc.FilePath = "Primitive"
	meshDesc.ShouldCull = true
	
	mesh:Initialize(meshDesc, material)
	
	-- Water	
	texDesc = TextureDesc.new()
	meshDesc = MeshDesc.new()
	
	trans = Transform.new()
	trans.Scale = Vector3f.new(15, 15, 15)
	
	gameObject = Scene:InstantiateGameObject("Water", trans, false)
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(0, "Water.vs")
	material:AddShader(0, "Water.fs")
	
	texDesc.RegisterIndex = 0
	texDesc.UniformName = "uNoise"
	texDesc.FilePath = "Assets/Textures/Noise.tga"
	material:AddTexture(texDesc)
	
	texDesc.RegisterIndex = 1
	texDesc.UniformName = "uColor"
	texDesc.FilePath = "Assets/Textures/Colormap.tga"
	material:AddTexture(texDesc)
	
	-- Primitives is static
	Primitives.GetPlaneTri(meshDesc, 200, 0, 200)
	meshDesc.HasIndices = true
	meshDesc.FilePath = "Primitive 200 200"
	
	mesh:Initialize(meshDesc, material)
	
	trans = Transform.new()
	trans.Position = Vector3f.new(0, -15, 0)
	trans.Scale = Vector3f.new(15, 1, 15)
	
	gameObject = Scene:InstantiateGameObject("GROUND FOR SHADOW MAPPING", trans, false)
	mesh = gameObject:AddMeshRenderer()
	material = mesh:SetNewMaterial()
	
	material:AddShader(1, "Shadow/Depth.vs")
	material:AddShader(1, "Shadow/Depth.fs")
	
	-- Primitives is static
	Primitives.GetPlaneTri(meshDesc, 1, 1, 1)
	meshDesc.HasIndices = true
	meshDesc.ShouldCull = false
	meshDesc.FilePath = "Primitive 1 1"
	
	mesh:Initialize(meshDesc, material)
	
end

function Update()

end